extensions/pkg/controller/worker/genericactuator/actuator.go

```diff
@@ -383,3 +384,38 @@ func CleanupLeakedClusterRoles(ctx context.Context, c client.Client, providerNam
 
 	return flow.Parallel(fns...)(ctx)
 }
+
+// runMachineDependencyOperation is a wrapper to run a machine dependency operation
+// and to persist the returned status object in .status.providerStatus of the Worker resource.
+func (a *genericActuator) runMachineDependencyOperation(ctx context.Context, worker *extensionsv1alpha1.Worker, operationFn func(context.Context) (runtime.Object, error)) error {
+	// Run the machine dependency operation.
+	workerProviderStatus, err := operationFn(ctx)
+
```

Comment:
Hmm I guess the issue is that `machine images` and now also `machine dependencies` are stored in the provider status together. The provider status in the worker status is of type `*runtime.RawExtension` and I guess therefore the contents of the provider status are a blackbox for the `genericactuator`, right?
If yes, we can't really distinguish the contents in the provider status and it is in the responsibility of the provider extensions.

So if we want to separate the things, to make updates for `machine images` and `machine dependencies` independently, then we could consider to have dedicated fields of type `*runtime.RawExtension` in the worker status for the `machine images` and `machine dependencies`. The results returned by `GetMachineImages` would be stored in `.status.machineImages` and the results returned by `DeployMachineDependencies` and `DeployMachineDependencies` would be stored in `.status.machineDependencies`.

So not really have another idea to prevent information lost if an extension does not behave correctly and return an invalid provider status.

Reply:
No, I’m against multiple fields. We should not let implementation details drive the API design.
What about introducing a `UpdateProviderStatus` method in the `WorkerDelegate` interface which is called once after the machine classes/machine deployments/machine dependencies were deployed (but before we wait for the machine deployments to be available)? The concrete implementations in the extensions can then construct this status as necessary.

Reply:
Hmm sounds promising. Let me check. 👍

Reply:
I played a little bit around with an `UpdateProviderStatus` of the `WorkerDelegate` interface.
I guess on issue is that we would need to call the `UpdateProviderStatus` method multiple times.

In the reconciliation flow we would need to call it after `DeployMachineDependencies` + `GetMachineImages` and once more after `CleanupMachineDependencies`. In the deletion flow just once after `CleanupMachineDependencies`.


The methods `DeployMachineDependencies/CleanupMachineDependencies` and `GetMachineImages` would return different substatus of the worker provider status. Then we could extend the `WorkerDelegate` interface with method like this:
```
UpdateProviderStatus(runtime.Object, runtime.Object) error
```
So the first `runtime.Object` parameter is for the machine image substatus and the second one for the machine dependencies substatus.

I guess this would work, but it would bring some assumption into the `genericactuator` how the worker provider status should be structured. I'm not sure if we want that. IMHO the worker provider status should be a blackbox for the `genericactuator` and only the respective provider-extension should decide how the status is structured.
But in this case the provider-extension need to make sure that the status is generated correctly and that no existing fields will be overridden.

So what is your option on that?

Reply:
I agree to your statement that we don't want to bring these assumptions about the provider status structure into the generic actuator. 
I would rather propose to replace the `GetMachineImages()` function in the `WorkerDelegate` interface with `UpdateMachineImagesStatus()`, and introduce `UpdateMachineDependenciesStatus()`. The actual implementation of the `WorkerDelegate` can then update the `.status` of the `Worker` itself.

WDYT?

Reply:
Yes, I agree. The provider extension only should be responsible for the worker provider status.

So lets replace the `GetMachineImages(context.Context) (runtime.Object, error)` method with a `UpdateMachineImageStatus(context.Context) error` method. The difference here is that `UpdateMachineImageStatus` need to update the worker provider status itself.

I would like to keep the `DeployMachineDependencies(context.Context) error /CleanupMachineDependencies(context.Context) error` methods (without returning a `runtime.Object` as they also update the worker provider status on their own) instead of one `UpdateMachineDependenciesStatus(context.Context) error` . They are called at different times in the flow and we need to do different things in those functions. We could probably also do everything in one method but this would make the implementations more complex.

WDYT? If you agree I would update the PR accordingly.

Reply:
Yes, I think that sounds reasonable :)